### PR TITLE
Skip generate lifecycle methods if they already generated

### PR DESCRIPTION
**Pull request by aivus**
This happens when I use one lifecycle method in prePersist and preUpdate together. Like

``` yml
    lifecycleCallbacks:
        prePersist: [setCreatedAtValue, setUpdatedAtValue]
        preUpdate: [setUpdatedAtValue]
```

Generator generate two methods with the same names.

@Ocramius, what do you think what is better way to write test for this?
Should I test generateEntityLifecycleCallbackMethods directly?

### CODE REVIEW

**Comment by DHager** [comment]
It sounds to me like it ought to generate one method with no form of type-hinting because it needs to accept multiple kinds of arguments.

**Comment by aivus** [comment]
@DHager, can you explain your answer?

**Comment by DHager** [comment]
@aivus : Sorry, I had assumed that the generation code already had type-hints for arguments, [but I guess it doesn't](https://github.com/doctrine/doctrine2/blob/b923c937e2194ae38c88e31cfcea8d06b58603fc/lib/Doctrine/ORM/Tools/EntityGenerator.php#L289). I was thinking of a scenario where a user had specified (incompatibly) that they wanted both:

    public function setUpdatedAtValue(PreUpdateEventArgs $args){...}
    public function setUpdatedAtValue(LifecycleEventArgs $event){...}

In which case you wouldn't have been able to "just use the first".

**Comment by aivus** [comment]
@Ocramius what do you think about this PR?

**Comment by Ocramius** [comment]
I don't see why you'd want to generate lifecycle callbacks on entities in first place?

**Comment by aivus** [comment]
@Ocramius because second lifecycle callback can have name like already generated callback.
In my example will be generated **two** methods `setUpdatedAtValue`.

**Comment by aivus** [comment]
@DHager but now this callbacks have equal signature.

**Comment by Ocramius** [comment]
> @Ocramius because second lifecycle callback can have name like already generated callback.

I still don't understand why you'd generate a callback anyway

**Comment by aivus** [comment]
Do you mean config from first post?
I need to update `updated_at` column and do some logic when record was created and updated.

**Comment by Ocramius** [comment]
Let me rephrase: why are you generating such a method, instead of writing it?

**Comment by aivus** [comment]
Generating entity getters/settes and generate lifecycle methods too.
Sometimes it's easy to generate basic set of getters/setters instead of write it manually.

**Comment by Ocramius** [comment]
Yes, but lifecycle callbacks are logic containers, and will contain custom logic anyway, so I don't see why you'd generate them anyway (same question as above). What's the advantage of having an empty pre-generated method?

**Comment by aivus** [comment]
This methods generated when I generate base set too.
And I should remove method duplicates because I get exception.

**Comment by Ocramius** [comment]
@aivus if it's a bug, then add a test case first, please

**Comment by aivus** [comment]
Ok, sure

**Comment by aivus** [comment]
@Ocramius, is it correct to have one method with few lifecycle callback annotations?
Like
```php
/*
 * @PreUpdate
 * @PrePersist
 */
public function setUpdatedAt()
{
    // 
}
```

**Comment by Ocramius** [comment]
That is supposed to work, yes

**Comment by aivus** [comment]
@Ocramius PR is ready. Please, review when you can.

**Comment by aivus** [comment]
@Ocramius ping

**Comment by aivus** [comment]
@Ocramius can you review this?

File: tests/Doctrine/Tests/ORM/Tools/EntityGeneratorTest.php [hunk 1]
```diff
@@ -82,6 +82,8 @@ public function generateBookEntityFixture(array $embeddedClasses = array())
         ));
         $metadata->addLifecycleCallback('loading', 'postLoad');
         $metadata->addLifecycleCallback('willBeRemoved', 'preRemove');
+        $metadata->addLifecycleCallback('updateDate', 'preUpdate');
+        $metadata->addLifecycleCallback('updateDate', 'prePersist');
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Tools/EntityGeneratorTest.php]: To be moved to a separate test method that verifies method de-duplication explicitly

File: lib/Doctrine/ORM/Tools/EntityGenerator.php [hunk 2]
```diff
@@ -1395,22 +1400,30 @@ protected function generateEntityStubMethod(ClassMetadataInfo $metadata, $type,
...
+    protected function generateLifecycleCallbackMethod($lifecycleNames, $methodName, $metadata)
     {
         if ($this->hasMethod($methodName, $metadata)) {
             return '';
         }
-        $this->staticReflection[$metadata->name]['methods'][] = $methodName;
+
+        $this->staticReflection[$metadata->name]['methods'][] = strtolower($methodName);
+
+        $lifecycleReplacements = array_map(
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: Do the `implode()` directly here as well

File: lib/Doctrine/ORM/Tools/EntityGenerator.php [hunk 3]
```diff
@@ -1395,22 +1400,30 @@ protected function generateEntityStubMethod(ClassMetadataInfo $metadata, $type,
...
     /**
-     * @param string            $name
+     * @param string[]          $lifecycleNames
      * @param string            $methodName
      * @param ClassMetadataInfo $metadata
      *
      * @return string
      */
-    protected function generateLifecycleCallbackMethod($name, $methodName, $metadata)
+    protected function generateLifecycleCallbackMethod($lifecycleNames, $methodName, $metadata)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: This is a BC break, since you are changing the method signature

File: lib/Doctrine/ORM/Tools/EntityGenerator.php [hunk 4]
```diff
@@ -1252,20 +1252,25 @@ protected function isAssociationIsNullable($associationMapping)
     protected function generateEntityLifecycleCallbackMethods(ClassMetadataInfo $metadata)
     {
         if (isset($metadata->lifecycleCallbacks) && $metadata->lifecycleCallbacks) {
+            $methodsCode = array();
+
             $methods = array();
+            foreach ($metadata->lifecycleCallbacks as $lifecycleName => $methodNames) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: Empty line required between logical blocks

Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: Consider moving this block into a private method

File: lib/Doctrine/ORM/Tools/EntityGenerator.php (hunk 3 above)
Comment (aivus) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: What do you think about revert signature and check `$name` on `array`?
Like
```php
if (!is_array($name)) {
    $lifecycleNames = [$name];
}
```

Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/EntityGenerator.php]: No, that involves even more chaos later on. I suggest not reusing this method for this new purpose at all...

**Comment by aivus** [comment]
I've updated PR.
Hope it's ok now.

Can I use short array syntax for new code?

**Comment by aivus** [comment]
@Ocramius ping

**Comment by aivus** [comment]
@Ocramius could you review this PR?

**Comment by aivus** [comment]
@Ocramius ping

**Comment by aivus** [comment]
Closed due to no feedback from doctrine maintainers
